README.md
# ziiir.com open-source web communication platform

[![Build Status][action-image]][action-url]
[![codecov][codecov-image]][codecov-url]
[![License][license-image]][license-url]

[action-image]: https://github.com/mikbry/ziiircom/workflows/Tests%20and%20Deploy/badge.svg
[action-url]: https://mikbry.github.io/ziiircom/
[codecov-image]: https://codecov.io/gh/mikbry/ziiircom/branch/master/graph/badge.svg?token=Y6MAYduEEz
[codecov-url]: https://codecov.io/gh/mikbry/ziiircom
[License-url]:https://github.com/mikbry/ziiircom/blob/master/LICENSE
[license-image]: https://img.shields.io/github/license/mikbry/ziiircom


## RoadMap
# v0.1
- [x] code base
- [x] components: minimal
- [x] messenger : minimal
- [ ] messenger-react : minimal
- [ ] bot : parse conversation.json
- [ ] admin : create config.json test conversation.json, export bundle


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# 0.1.0 (2020-07-01)


### Bug Fixes

* default config empty ([#99](https://github.com/mikbry/ziiircom/issues/99)) ([9da0fb7](https://github.com/mikbry/ziiircom/commit/9da0fb74788b371718ee0b00a06ddf516adb9c56))
* load messageHook without config ([#97](https://github.com/mikbry/ziiircom/issues/97)) ([2115ce9](https://github.com/mikbry/ziiircom/commit/2115ce9281a8e32a03e8764f19e514daf374426f))


### Features

* add #reset command ([#61](https://github.com/mikbry/ziiircom/issues/61)) ([0437d7a](https://github.com/mikbry/ziiircom/commit/0437d7a5f7834e6cc5a8908a5ed3a70c7f205e3a))
* add input in messenger [#43](https://github.com/mikbry/ziiircom/issues/43) ([#45](https://github.com/mikbry/ziiircom/issues/45)) ([69590d6](https://github.com/mikbry/ziiircom/commit/69590d6a177a6acc95ea202e886b45537197804a))
* add message.created_time ([#62](https://github.com/mikbry/ziiircom/issues/62)) ([7767507](https://github.com/mikbry/ziiircom/commit/77675070890e0655e4cc6f493ad0f2c6362f005e))
* add theme config support [#23](https://github.com/mikbry/ziiircom/issues/23) ([#42](https://github.com/mikbry/ziiircom/issues/42)) ([9f7c33b](https://github.com/mikbry/ziiircom/commit/9f7c33bb9bd66d17c3f960de6b1f3ed9bc8792f3))
* common package created ([#69](https://github.com/mikbry/ziiircom/issues/69)) ([9c9e2bb](https://github.com/mikbry/ziiircom/commit/9c9e2bb3fa7ace5aec54c28805d678c022d232f6))
* component create Button [#28](https://github.com/mikbry/ziiircom/issues/28) ([#35](https://github.com/mikbry/ziiircom/issues/35)) ([4b958cc](https://github.com/mikbry/ziiircom/commit/4b958ccdd41892f758ad1bd88c1a4f5cc65d388f))
* component FAButton added [#21](https://github.com/mikbry/ziiircom/issues/21) ([#22](https://github.com/mikbry/ziiircom/issues/22)) ([a943a35](https://github.com/mikbry/ziiircom/commit/a943a35a59747f6dafea16d915ffdd38491a1e39))
* components add Input ([#36](https://github.com/mikbry/ziiircom/issues/36)) ([d9a4a17](https://github.com/mikbry/ziiircom/commit/d9a4a17b86849ab8f269f21d8ec634371696f7c8))
* create messaging package ([#125](https://github.com/mikbry/ziiircom/issues/125)) ([01b77aa](https://github.com/mikbry/ziiircom/commit/01b77aabdc063cde706e6f8d123e6a71a7464365))
* dialog add any input ([#127](https://github.com/mikbry/ziiircom/issues/127)) ([531c56f](https://github.com/mikbry/ziiircom/commit/531c56f827c0748eb3b3af95170e8ee757ce67b5))
* dialog buildOutput ([#126](https://github.com/mikbry/ziiircom/issues/126)) ([706cf5d](https://github.com/mikbry/ziiircom/commit/706cf5de37529faab0ce8bec1ad69b8de6743f28))
* dialog conditions ([#136](https://github.com/mikbry/ziiircom/issues/136)) ([b7eaf1e](https://github.com/mikbry/ziiircom/commit/b7eaf1ea206293c379e022540fbc39afbe258cb9))
* dialog context ([#134](https://github.com/mikbry/ziiircom/issues/134)) ([1e55620](https://github.com/mikbry/ziiircom/commit/1e55620a31951d0f1a951f282914a6dad362c088))
* dialog create package ([#118](https://github.com/mikbry/ziiircom/issues/118)) ([3403add](https://github.com/mikbry/ziiircom/commit/3403add706b39e81adc29a53d1036b178f50fdf5))
* dialog variables templating ([#135](https://github.com/mikbry/ziiircom/issues/135)) ([7650d56](https://github.com/mikbry/ziiircom/commit/7650d564dd7a1e4e43189140d7add408ef86fc2b))
* display friendly date in message ([#70](https://github.com/mikbry/ziiircom/issues/70)) ([18de2a4](https://github.com/mikbry/ziiircom/commit/18de2a4fb98ea225a12b171bb60ca7855c4e4dd0))
* fetch intents from src ([#122](https://github.com/mikbry/ziiircom/issues/122)) ([c3c3131](https://github.com/mikbry/ziiircom/commit/c3c313129cb9d2e33b8198474d6402b08a4a9d5a))
* global init ([#3](https://github.com/mikbry/ziiircom/issues/3)) ([44e4f2d](https://github.com/mikbry/ziiircom/commit/44e4f2dae1c2f3e76306f9d9b083bb4ff2a8cbce))
* handle button action ([#142](https://github.com/mikbry/ziiircom/issues/142)) ([c937eb6](https://github.com/mikbry/ziiircom/commit/c937eb6309d24ba2bf6e7fe817c70f2db343a393))
* handle message from input ([#49](https://github.com/mikbry/ziiircom/issues/49)) ([90d078d](https://github.com/mikbry/ziiircom/commit/90d078db73d6c848e484aab579fa0cd3e4b4c8f5))

main.345daf78f05eed963256.js
!function(e){var t={};function r(a){if(t[a])return t[a].exports;var n=t[a]={i:a,l:!1,exports:{}};return e[a].call(n.exports,n,n.exports,r),n.l=!0,n.exports}r.m=e,r.c=t,r.d=function(e,t,a){r.o(e,t)||Object.defineProperty(e,t,{enumerable:!0,get:a})},r.r=function(e){"undefined"!=typeof Symbol&&Symbol.toStringTag&&Object.defineProperty(e,Symbol.toStringTag,{value:"Module"}),Object.defineProperty(e,"__esModule",{value:!0})},r.t=function(e,t){if(1&t&&(e=r(e)),8&t)return e;if(4&t&&"object"==typeof e&&e&&e.__esModule)return e;var a=Object.create(null);if(r.r(a),Object.defineProperty(a,"default",{enumerable:!0,value:e}),2&t&&"string"!=typeof e)for(var n in e)r.d(a,n,function(t){return e[t]}.bind(null,n));return a},r.n=function(e){var t=e&&e.__esModule?function(){return e.default}:function(){return e};return r.d(t,"a",t),t},r.o=function(e,t){return Object.prototype.hasOwnProperty.call(e,t)},r.p="",r(r.s=10)}([function(e,t,r){"use strict";Object.defineProperty(t,"__esModule",{value:!0}),t.default=t.setup=void 0;const a={};t.setup=({html:e,styled:t,createElement:r})=>(a.html||(a.html=e,a.styled=t,a.createElement=r,Object.freeze(a)),a);var n=a;t.default=n},function(e,t,r){"use strict";Object.defineProperty(t,"__esModule",{value:!0}),t.default=void 0;var a={palette:{primary:"palevioletred",secondary:"#70DBB8",background:"smokewhite",surface:"papayawhip",onPrimary:"white",onSecondary:"white",onBackground:"dark",onSurface:"gray",disabledText:"lightGrey"},fonts:[],font:{size:"16px",style:"normal",weight:"400"},spacing:8,radius:12,smallRadius:4,shadow:"rgba(0,0,0,0.2) 0px 2px 4px -1px,rgba(0,0,0,0.14) 0px 4px 5px 0px,rgba(0,0,0,0.12) 0px 1px 10px 0px",messenger:{width:"400px",height:"600px"}};t.default=a},function(e,t,r){"use strict";Object.defineProperty(t,"__esModule",{value:!0}),Object.defineProperty(t,"friendlyDate",{enumerable:!0,get:function(){return n.friendlyDate}}),Object.defineProperty(t,"deepCopy",{enumerable:!0,get:function(){return o.default}});var a,n=r(23),o=(a=r(24))&&a.__esModule?a:{default:a}},function(e,t,r){"use strict";Object.defineProperty(t,"__esModule",{value:!0}),t.useRef=t.render=t.createElement=t.html=void 0;var a=r(2),n=r(7);t.html=()=>{};t.createElement=(e,t,...r)=>{let n;e.element?(n=(0,a.deepCopy)({},e),n.props=(0,a.deepCopy)({},n.props,t)):(n={props:{}},n.element=e,t&&(n.props=t));let o=r;return 1===r.length&&Array.isArray(r[0])&&([o]=r),n.children=o.filter(e=>void 0!==e),n};const o=(e,t,r={},s)=>{if(!e)return;if("string"==typeof e){return void(t.innerText=e)}let{children:l}=e;if("string"==typeof e.element){const i=document.createElement(e.element);if(e.props&&Object.keys(e.props).forEach(t=>{const r=e.props[t],a=typeof r,n="string"===a||"number"===a||"boolean"===a;"ref"===t?r.setCurrent(i):"className"===t?i.className=r:"function"===a?"on"===t.substring(0,2)&&i.addEventListener(t.substring(2).toLowerCase(),r):n&&i.setAttribute(t,r)}),e.styled){const t={...e.defaultProps},o=r.theme||{},s=(0,a.deepCopy)(t,e.props,{theme:o});let l=(0,n.genClassRules)(e.styled,s,t);e.props.className&&(l+=" "+e.props.className),i.className=l}l&&l.forEach(e=>{o(e,i,r)}),s&&s.before?t.insertBefore(i,s.before):t.appendChild(i)}else if("function"==typeof e){const a=e({...e.props,children:l,styled:e.styled});o(a,t,r),l=null}else if(e.element&&"function"==typeof e.element){const a=e.element({...e.props,children:l});e.styled&&(a.styled=[...e.styled,...a.styled]),o(a,t,r),l=null}};t.render=o;t.useRef=e=>{const t={current:e,setCurrent:e=>{t.current=e}};return t}},function(e,t,r){"use strict";Object.defineProperty(t,"__esModule",{value:!0}),t.default=async function(){const e=await Promise.resolve().then(()=>o(r(12))),t=await Promise.resolve().then(()=>o(r(5))),a=await Promise.resolve().then(()=>o(r(27)));return{Messenger:e.default,Message:t.default,FAButton:a.default}},Object.defineProperty(t,"Interface",{enumerable:!0,get:function(){return a.default}}),Object.defineProperty(t,"setup",{enumerable:!0,get:function(){return a.setup}});var a=o(r(0));function n(){if("function"!=typeof WeakMap)return null;var e=new WeakMap;return n=function(){return e},e}function o(e){if(e&&e.__esModule)return e;if(null===e||"object"!=typeof e&&"function"!=typeof e)return{default:e};var t=n();if(t&&t.has(e))return t.get(e);var r={},a=Object.defineProperty&&Object.getOwnPropertyDescriptor;for(var o in e)if(Object.prototype.hasOwnProperty.call(e,o)){var s=a?Object.getOwnPropertyDescriptor(e,o):null;s&&(s.get||s.set)?Object.defineProperty(r,o,s):r[o]=e[o]}return r.default=e,t&&t.set(e,r),r}},function(e,t,r){"use strict";var a;Object.defineProperty(t,"__esModule",{value:!0}),t.default=void 0;var n=((a=r(22))&&a.__esModule?a:{default:a}).default;t.default=n},function(e,t,r){"use strict";Object.defineProperty(t,"__esModule",{value:!0}),t.default=void 0;var a,n=(a=r(4))&&a.__esModule?a:{default:a};let o;var s=async()=>(o||(o=await(0,n.default)()),o);t.default=s},function(e,t,r){"use strict";let a;Object.defineProperty(t,"__esModule",{value:!0}),t.genClassRules=t.genStyle=void 0;const n="abcdefghijklmnopqrstuvwxyz",o=(e,t)=>{let r="";return e&&Array.isArray(e.styles)&&e.styles.forEach((a,n)=>{r+=a,e.parameters[n]&&(r+=e.parameters[n](t))}),r};t.genStyle=o;const s=(e,t)=>{const r=e;if(0===t.trim().length)return;let a=0,n=t.indexOf("@",a);for(;n>-1;){r.main+=t.substring(a,n).trim();const e=t.indexOf("@",n+1);a=-1!==e?e:t.lastIndexOf("}")+1,r.medias.push(t.substring(n,a)),n=e}r.main+=t.substring(a).trim()};t.genClassRules=(e,t,r={})=>{const l=(()=>{if(!a){const e=document.createElement("style");e.type="text/css",e.title="MainCSS",document.getElementsByTagName("head")[0].appendChild(e),a=e.sheet}return a})();let i="";return e.forEach(e=>{const a=r.selector||[...Array(5)].map(()=>n[~~(Math.random()*n.length)]).join(""),u={main:"",sub:[],medias:[]};if(((e,t,r)=>{const a=o(t,r);let n=0,l=a.indexOf("&",n);if(-1===l)s(e,a);else{for(;l>-1;)s(e,a.substring(n,l)),n=a.indexOf("}",l+2),e.sub.push(a.substring(l+1,n+1)),n+=1,l=a.indexOf("&",n);s(e,a.substring(n))}})(u,e,t),u.main&&u.main.length>0){let e=`.${a} { ${u.main} }`;l.insertRule(e),u.sub.forEach(t=>{e=`.${a}${t}`,l.insertRule(e)}),u.medias.forEach(e=>{l.insertRule(e)}),i+=a+" "}}),i.trim()}},function(e,t,r){"use strict";Object.defineProperty(t,"__esModule",{value:!0}),Object.defineProperty(t,"genStyle",{enumerable:!0,get:function(){return n.genStyle}}),t.styled=void 0;var a=r(3),n=r(7);t.styled=(e,t,...r)=>(n,...o)=>{const s=e,l=(0,a.createElement)(s,t,...r);return l.styled?l.styled.unshift({styles:n,parameters:o}):l.styled=[{styles:n,parameters:o}],l}},function(e,t,r){"use strict";let a;Object.defineProperty(t,"__esModule",{value:!0}),t.useSelector=t.default=void 0;t.default=e=>(a={...e},a);t.useSelector=e=>e({...a})},function(e,t,r){"use strict";Object.defineProperty(t,"__esModule",{value:!0}),t.default=t.defaultClient=void 0;var a,n=(a=r(11))&&a.__esModule?a:{default:a};function o(){if("function"!=typeof WeakMap)return null;var e=new WeakMap;return o=function(){return e},e}function s(e){if(e&&e.__esModule)return e;if(null===e||"object"!=typeof e&&"function"!=typeof e)return{default:e};var t=o();if(t&&t.has(e))return t.get(e);var r={},a=Object.defineProperty&&Object.getOwnPropertyDescriptor;for(var n in e)if(Object.prototype.hasOwnProperty.call(e,n)){var s=a?Object.getOwnPropertyDescriptor(e,n):null;s&&(s.get||s.set)?Object.defineProperty(r,n,s):r[n]=e[n]}return r.default=e,t&&t.set(e,r),r}t.defaultClient=async(e,t,a={messenger:{}})=>{let o,l;try{o=await Promise.resolve().then(()=>s(r(30))),o=o.default}catch(e){}o||(o={});try{l=(await Promise.resolve().then(()=>s(r(31)))).default}catch(e){}const i={...o,...a};return(0,n.default)({state:i,messageHook:l,root:e,messageListener:t})};var l=n.default;t.default=l},function(e,t,r){"use strict";Object.defineProperty(t,"__esModule",{value:!0}),t.default=void 0;var a=r(4),n=u(r(6)),o=r(3),s=r(8),l=u(r(9)),i=u(r(29));function u(e){return e&&e.__esModule?e:{default:e}}var d=async({state:e,messageHook:t,root:r,messageListener:u=(({type:e,message:t})=>({type:e,message:t}))})=>{(0,a.setup)({html:o.html,createElement:o.createElement,styled:s.styled});const d=await(0,n.default)(),f=(0,l.default)(e),c=async({type:e,message:t})=>{const r=document.getElementsByClassName("ziiir-conversation")[0];if("newMessage"===e){const e=(0,o.createElement)(d.Message,{key:t.created_time,createdtime:t.created_time,avatar:t.avatar,fromUser:"user"===t.from},t.text);let a;for(const e of r.children){const r=parseInt(e.getAttribute("created-time"),10);if(t.created_time<r){a={before:e};break}}(0,o.render)(e,r,void 0,a),r.scrollTop=r.scrollHeight}else if("resetMessages"===e)for(;r.firstChild;)r.firstChild.remove();u({type:e,message:t})},[p,m,y,g]=t(c),h=await p(),v=await(0,i.default)(h,e=>{if("#"===e.charAt(0))g(e);else{const t=m("user",e);y(t)}});return(0,o.render)(v,r,{...f}),[f,p,m,y,g,c]};t.default=d},function(e,t,r){"use strict";var a;Object.defineProperty(t,"__esModule",{value:!0}),t.default=void 0;var n=((a=r(13))&&a.__esModule?a:{default:a}).default;t.default=n},function(e,t,r){"use strict";Object.defineProperty(t,"__esModule",{value:!0}),t.default=void 0;var a=f(r(0)),n=f(r(14)),o=f(r(16)),s=f(r(18)),l=f(r(20)),i=f(r(5)),u=f(r(25)),d=f(r(1));function f(e){return e&&e.__esModule?e:{default:e}}const c=a.default.styled(n.default)`
  color: ${e=>e.theme.palette.onSurface};
  font-family: sans-serif;
  font-size: 16px;
  font-style: normal;
  font-weight: 400;
  line-height: 1.1;
  overflow-wrap: break-word;

  padding: 0px;
  width: 400px;
  height: 600px;
  display: flex;
  border-radius: 12px;
  flex-direction: column;
  box-shadow: rgba(0, 0, 0, 0.2) 0px 2px 4px -1px, rgba(0, 0, 0, 0.14) 0px 4px 5px 0px, rgba(0, 0, 0, 0.12) 0px 1px 10px 0px;

  &.isexpanded .ziiir-conversation {
    animation: 0.675s cubic-bezier(0.4, 0, 0.2, 1) 0.45s 1 normal both running enterConversation;
    animation-delay: 0.45s;
    animation-fill-mode: both;
  }

  @keyframes enterConversation {
    0% {
      height: 0;
      padding: 0;
      opacity: 0;
    }
  
    50% {
      opacity: 0;
    }
  
    100% {
      height: 100%;
      padding: 16px;
      opacity: 1;
    }
  }
`;c.defaultProps={theme:d.default};const p=a.default.styled(o.default)`
  border-radius: 12px 12px 0 0;
`,m=a.default.styled(l.default)`
  border-radius: 0 0 12px 12px;
`,y=a.default.styled(s.default)`
  transform-origin: center bottom 0px;
  scroll-behavior: smooth;
`,g=a.default.styled(u.default)`
  width: 100%;
  padding: 4px;
  margin: 0;
  border: none;
`,h=a.default.createElement;var v=({isExpanded:e=!0,input:t={display:!0},header:r={},messages:a=[],onMessage:n,onClick:o})=>{const s=h(y,{isExpanded:e,className:"ziiir-conversation"},a.sort((e,t)=>e.createdtime>t.createdtime).map(e=>((e,t,r,a)=>h(i.default,{key:e,createdtime:e,avatar:a,fromUser:r},t))(e.createdtime,e.text,"user"===e.from,e.avatar)));let l="ziiir.com";const u=e=>{const t=e.target.value||"";"Enter"===e.key&&t.length>0&&(n&&n(t),e.target.value="")};return t.display&&(l=h(g,{className:"ziir-input",onKeyUp:u,placeholder:t.placeholder||"Your message"})),h(c,{className:e?"isexpanded":void 0,onClick:o},h(p,null,r.text||""),s,h(m,null,l))};t.default=v},function(e,t,r){"use strict";var a;Object.defineProperty(t,"__esModule",{value:!0}),t.default=void 0;var n=((a=r(15))&&a.__esModule?a:{default:a}).default;t.default=n},function(e,t,r){"use strict";Object.defineProperty(t,"__esModule",{value:!0}),t.default=void 0;var a=o(r(0)),n=o(r(1));function o(e){return e&&e.__esModule?e:{default:e}}const s=a.default.styled("div")`
  padding: 1.2em;
  background: ${e=>e.theme.palette.background};
  color: #292c45;
  font-family: Roboto Condensed,sans-serif;
  font-size: 16px;
  font-style: normal;
  font-weight: 400;
  line-height: 1.1;
  overflow-wrap: break-word;
`;s.defaultProps={theme:n.default};var l=s;t.default=l},function(e,t,r){"use strict";var a;Object.defineProperty(t,"__esModule",{value:!0}),t.default=void 0;var n=((a=r(17))&&a.__esModule?a:{default:a}).default;t.default=n},function(e,t,r){"use strict";Object.defineProperty(t,"__esModule",{value:!0}),t.default=void 0;var a=o(r(0)),n=o(r(1));function o(e){return e&&e.__esModule?e:{default:e}}const s=a.default.styled("header")`
  color: ${e=>e.theme.palette.onPrimary};
  font-size: 1.25em;
  background: ${e=>e.theme.palette.primary};
  min-height: 64px;
  padding-top: 1.2em;
  padding-left: 1.2em;
  padding-right: 0.3em;
  display: flex;
`;s.defaultProps={theme:n.default};var l=s;t.default=l},function(e,t,r){"use strict";var a;Object.defineProperty(t,"__esModule",{value:!0}),t.default=void 0;var n=((a=r(19))&&a.__esModule?a:{default:a}).default;t.default=n},function(e,t,r){"use strict";Object.defineProperty(t,"__esModule",{value:!0}),t.default=void 0;var a=o(r(0)),n=o(r(1));function o(e){return e&&e.__esModule?e:{default:e}}const s=a.default.styled("div")`
  padding: 0px;
  background: ${e=>e.theme.palette.background};
  display: flex;
  flex-direction: column;
  flex-grow: 1;
  overflow-y: auto;
  & > div {
  }
`;s.defaultProps={theme:n.default};var l=s;t.default=l},function(e,t,r){"use strict";var a;Object.defineProperty(t,"__esModule",{value:!0}),t.default=void 0;var n=((a=r(21))&&a.__esModule?a:{default:a}).default;t.default=n},function(e,t,r){"use strict";Object.defineProperty(t,"__esModule",{value:!0}),t.default=void 0;var a=o(r(0)),n=o(r(1));function o(e){return e&&e.__esModule?e:{default:e}}const s=a.default.styled("div")`
  color: ${e=>e.theme.palette.disabledText};
  font-size: 1em;
  line-height: 36px;
  min-height: 36px;
  padding-left: 16px;
  padding-right: 16px;
  display: flex;
  margin-top: auto;
  & > button {
    align-self: flex-end;
    line-height: 24px;
  }
`;s.defaultProps={theme:n.default};var l=s;t.default=l},function(e,t,r){"use strict";Object.defineProperty(t,"__esModule",{value:!0}),t.default=void 0;var a=r(2),n=s(r(0)),o=s(r(1));function s(e){return e&&e.__esModule?e:{default:e}}const l=n.default.styled("div")`
  display: flex;
  flex-direction: row;
  width: 100%;
  max-width: 80%;
  margin: 12px 0px 0px;

  cursor: default;

  margin-left: ${e=>e.fromUser?"auto":0};

  &:first-child {
    margin-top: 0px;
  }

  & > div {
    width: 100%;
    animation: 0.3s cubic-bezier(0, 0, 0.2, 1) 0s 1 normal both running ${e=>e.fromUser?"enterMessageFromUser":"enterMessageFromBot"};
  }
  & > div > p {
    background-color: ${e=>e.fromUser?e.theme.palette.secondary:e.theme.palette.surface};
    color: ${e=>e.fromUser?e.theme.palette.onSecondary:e.theme.palette.onSurface};
    box-shadow: rgba(0, 0, 0, 0.2) 0px 1px 5px 0px, rgba(0, 0, 0, 0.14) 0px 2px 2px 0px, rgba(0, 0, 0, 0.12) 0px 3px 1px -2px;
    margin-bottom: 4px;
    padding: 16px;
    border-radius: ${e=>e.fromUser?"12px 12px 4px 12px":"12px 12px 12px 4px"};
    overflow-wrap: break-word;
    line-height: 1.44;
  }
  & > div > span {
    text-align: ${e=>e.fromUser?"right":"left"};
    display: block;
    margin: 0px 16px;
    transition: all 0.25s cubic-bezier(0.4, 0, 0.2, 1) 0s;
    opacity: 0.5;
    font-size: 0.75rem;
    font-weight: 400;
    line-height: 1.66;
  }

  @keyframes enterMessageFromUser {
    0% {
      transform: translate( 0, 20px );
      opacity: 0;
    }
  
    100% {
      transform: translate( 0, 0 );
      opacity: 1;
    }
  }

  @keyframes enterMessageFromBot {
    0% {
      transform: translate( - $message-avatar-size, 0 );
      opacity: 0;
    }
  
    100% {
      transform: translate( 0, 0 );
      opacity: 1;
    }
  }
`;l.defaultProps={theme:o.default};const i=n.default.styled("img")`
  z-index: 3;
  flex-shrink: 0;
  width: 40px;
  height: 40px;
  margin-top: auto;
  margin-right: 8px;
  animation: 0.3s cubic-bezier(0.4, 0, 0.2, 1) 0s 1 normal none running enterAvatar;
  border-radius: 50%;

  @keyframes enterAvatar {
    0% {
      transform: scale( .5, .5 );
      opacity: 0;
    }
  
    30% {
      transform: scale( 1.25, 1.25 );
      opacity: 1;
    }
  
    80% {
      transform: scale( .9, .9 );
    }
  
    100% {
      transform: scale( 1, 1 );
    }
  }  
`,u=n.default.createElement;var d=({createdtime:e,avatar:t,fromUser:r=!0,children:n})=>{const o=(0,a.friendlyDate)(e);return u(l,{fromUser:r,"created-time":e},u("div",null,u("p",null,n),u("span",null,o)),t&&u(i,{src:t.src,alt:t.name}))};t.default=d},function(e,t,r){"use strict";Object.defineProperty(t,"__esModule",{value:!0}),t.friendlyDate=void 0;t.friendlyDate=(e,t)=>{let r=.001*(Date.now()-e);const a=Math.floor(r/31557600);r-=31557600*a;const n=Math.floor(r/2629800);r-=2629800*n;const o=Math.floor(r/86400);if(r-=86400*o,n||a||o>6)return new Date(e).toLocaleString(t);if(o>1)return new Date(e).toLocaleString(t);if(1===o)return"yesterday";const s=Math.floor(r/3600);if(r-=3600*s,s>1)return s+" hours ago";if(1===s)return s+" hour ago";const l=Math.floor(r/60);return r-=60*l,l>1?l+" mins ago":1===l?l+" min ago":r>10?r+" secs ago":"just now"}},function(e,t,r){"use strict";Object.defineProperty(t,"__esModule",{value:!0}),t.default=void 0;var a=(...e)=>{const t=[],r=(...e)=>{let a;return e.forEach(e=>{Array.isArray(e)?(a||(a=[]),e.forEach(e=>{a.push(r(e))})):e instanceof Date?a=new Date(e.getTime()):e&&"object"==typeof e?(t.push(e),a||(a={}),t.push(e),Object.keys(e).forEach(n=>{const o=e[n];-1!==t.indexOf(o)?a[n]=o:a[n]=r(a[n],o)}),t.pop()):null!=e&&(a=e)}),a};return r(...e)};t.default=a},function(e,t,r){"use strict";var a;Object.defineProperty(t,"__esModule",{value:!0}),t.default=void 0;var n=((a=r(26))&&a.__esModule?a:{default:a}).default;t.default=n},function(e,t,r){"use strict";Object.defineProperty(t,"__esModule",{value:!0}),t.default=void 0;var a=o(r(0)),n=o(r(1));function o(e){return e&&e.__esModule?e:{default:e}}const s=a.default.styled("input")`
  font-size: 1em;
  margin: 0.6em;
  padding: 0.3em 1em;
  border-radius: 3px;

  color: ${e=>e.theme.palette.secondary};
  border: 2px solid ${e=>e.theme.palette.secondary};
`;s.defaultProps={theme:n.default};var l=s;t.default=l},function(e,t,r){"use strict";var a;Object.defineProperty(t,"__esModule",{value:!0}),t.default=void 0;var n=((a=r(28))&&a.__esModule?a:{default:a}).default;t.default=n},function(e,t,r){"use strict";Object.defineProperty(t,"__esModule",{value:!0}),t.default=void 0;var a=o(r(0)),n=o(r(1));function o(e){return e&&e.__esModule?e:{default:e}}const s=a.default.styled("div")`
width:48px;
height:48px;
border-radius:100%;
color: ${e=>e.theme.palette.onSecondary};
font-size: 1.25em;
line-height: 64px;
background: ${e=>e.theme.palette.secondary};
background-image: ${e=>e.icon?`url("${e.icon}")`:`url("${((e,t="ffffff")=>{let r=e;"#"===r[0]&&(r=r.substring(1));let a=t;return"#"===a[0]&&(a=a.substring(1)),`data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' x='0px' y='0px'%0Awidth='64' height='64'%0AviewBox='0 0 172 172'%0Astyle=' fill:%23000000;'%3E%3Cg fill='none' fill-rule='nonzero' stroke='none' stroke-width='1' stroke-linecap='butt' stroke-linejoin='miter' stroke-miterlimit='10' stroke-dasharray='' stroke-dashoffset='0' font-family='none' font-weight='none' font-size='none' text-anchor='none' style='mix-blend-mode: normal'%3E%3Cpath d='M0,172v-172h172v172z' fill='none'%3E%3C/path%3E%3Cg%3E%3Cpath d='M138.64141,105.21563c0,6.08047 -4.90469,10.98516 -10.98516,10.98516h-83.3125c-6.08047,0 -10.98516,-4.90469 -10.98516,-10.98516v-48.50937c0,-6.08047 4.90469,-10.98516 10.98516,-10.98516h83.3125c6.08047,0 10.98516,4.90469 10.98516,10.98516z' fill='%23${a}'%3E%3C/path%3E%3Cpath d='M113.21094,130.98203c-0.87344,0.87344 -2.25078,0.87344 -3.12422,0l-10.51484,-14.88203c-0.87344,-0.87344 -0.87344,-2.25078 0,-3.12422h24.1875c0.87344,0.87344 0.87344,2.25078 0,3.12422z' fill='%23${a}'%3E%3C/path%3E%3Cg fill='%23${r}'%3E%3Cpath d='M122.58359,64.36562c0,2.21719 -1.78047,3.99766 -3.99766,3.99766h-65.44063c-2.21719,0 -3.99766,-1.78047 -3.99766,-3.99766v0c0,-2.21719 1.78047,-3.99766 3.99766,-3.99766h65.47422c2.18359,0.03359 3.96406,1.81406 3.96406,3.99766zM122.58359,80.35625c0,2.21719 -1.78047,3.99766 -3.99766,3.99766h-65.44063c-2.21719,0 -3.99766,-1.78047 -3.99766,-3.99766v0c0,-2.21719 1.78047,-3.99766 3.99766,-3.99766h65.47422c2.18359,0 3.96406,1.78047 3.96406,3.99766zM122.58359,96.31328c0,2.21719 -1.78047,3.99766 -3.99766,3.99766h-65.44063c-2.21719,0 -3.99766,-1.78047 -3.99766,-3.99766v0c0,-2.21719 1.78047,-3.99766 3.99766,-3.99766h65.47422c2.18359,0 3.96406,1.78047 3.96406,3.99766z'%3E%3C/path%3E%3C/g%3E%3C/g%3E%3C/g%3E%3C/svg%3E%0A`})(e.theme.palette.secondary,e.theme.palette.bubbleColor)}")`};
background-repeat:no-repeat;
background-position:50%;
background-size:cover;
box-shadow:0 7px 8px -4px rgba(0,0,0,.2),0 12px 17px 2px rgba(0,0,0,.14),0 5px 22px 4px rgba(0,0,0,.12);
cursor:pointer;
`;s.defaultProps={theme:n.default};var l=s;t.default=l},function(e,t,r){"use strict";Object.defineProperty(t,"__esModule",{value:!0}),t.default=void 0;var a,n=(a=r(6))&&a.__esModule?a:{default:a},o=r(8),s=r(3),l=r(9);var i=async(e,t)=>{const[r,a]=(0,l.useSelector)(e=>[e.messenger.isOpen,e.messenger.header]);let i=r;const u=(0,s.useRef)(null),d=()=>{i=!i;const e=u.current;e.classList.toggle("isclosed"),e.classList.toggle("isopen")},f=await(0,n.default)(),c=(0,o.styled)(f.Messenger,{onMessage:e=>{t(e)},onClick:e=>{e.stopPropagation()},header:a})`
    width: 100%;
    max-width: 600px;
    height: 80%;
  `,p=(0,o.styled)("div",{ref:u,className:"ziiircom-messenger is"+(i?"open":"closed"),onClick:d},c)`
    display: flex;
    position: fixed;
    top: 0;
    left: 0;
    width: 100vw;
    height: 100vh;
    z-index: 3;
    flex-direction: column;
    flex-shrink: 0;
    align-items: center;
    justify-content: center;

    &.isopen {
      display: flex;
    }

    &.isclosed {
      display: none;
    }
  `,m=(0,o.styled)(f.FAButton,{className:"ziiircom-messenger-fab",onClick:d})`
    position: absolute;
    right: 32px;
    bottom: 32px;
    z-index: 9;
  `;return(0,o.styled)("div",{className:"ziiircom-messenger-frame"},p,m)`
    position: fixed;
    top: 0;
    left: 0;
    width: 100vw;
    height: 100vh;
    margin: 0;
    padding: 0;
  `};t.default=i},function(e){e.exports=JSON.parse("{}")},function(e,t,r){"use strict";Object.defineProperty(t,"__esModule",{value:!0}),t.default=void 0;var a=r(2);var n=e=>{const t=[],r=(e,t)=>({from:e,text:t,created_time:Date.now()});return[async()=>(0,a.deepCopy)(t),r,async n=>{t.push((0,a.deepCopy)(n)),await e({type:"newMessage",message:(0,a.deepCopy)(n)});const o=r("bot","You say : "+n.text);t.push(o),await e({type:"newMessage",message:(0,a.deepCopy)(o)})},async t=>{e("#reset"===t?{type:"resetMessages"}:{type:"unknownCommand"})}]};t.default=n}]);
//# sourceMappingURL=main.345daf78f05eed963256.js.map